Zephyr v3.3.0 xt

xakep-amatop opened the pull request:
**NOT FOR MERGE
CONTAINS ONLY XEN DOM0 RELATED CHANGES
IT IS ONLY FOR PRELIMINARY REVIEW BY XEN-TROOPS TEAM BEFORE OPENING PULL FOR ZEPHYR**

firscity committed:
xenvm: dom0: update Xen public headers for running Zephyr as Dom0

Xen public headers are added to Zephyr kernel partially, so we need
to update them for new features implementation.

Further implementation of Xen domains configuration and memory
management requires dedicated Xen public headers inside Zephyr kernel.
It will be used for add Zephyr OS Xen Domain-0 funtionalities.

Update existent and add new required public headers from Xen 4.17.0
release.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

firscity committed:
xen: change HVM functions signature to run it for other domains

This commit adds possibility to call hypervisor HVM parameter functions
for specified domain (instead of only DOMID_SELF). It is needed for
configuring domains, that were created from Zephyr control domain.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

firscity committed:
xenvm: events: add event channel allocation for domain-0

This commit adds interface for evtchn allocation that can be used by
privilaged domain. Domain 0 can specify both dom and remote_dom
parameters for hypercall, where in others domains dom should be always
DOMID_SELF. It is needed for creating pre-defined channels during
domain setup in Zephyr Dom0.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
soc: arm64: xenvm: move XEN_INITIAL_DOMAIN config to Xen driver

Untie XEN_INITIAL_DOMAIN from XenVM SoC configuration, it is
necessary for running Xen dom-0 on different boards. Move it from
XenVM SoC configuration to Xen driver. Set XEN_INITIAL_DOMAIN by
default to n in order to prevent issues with existing setups,
which may use XenVM as DomU.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
dts: bindings: xen: add xen,xen.yaml file

Add yaml file for 'xen,xen', because without it an appropriate
'CONFIG_DT_HAS_XEN_XEN_ENABLED' isn't generated.

Remove xen,xen-4.15.yaml at all, because it isn't necessary to have
yaml for some specific Xen version.

Add subdirectory of 'xen' when we have Xen Hypervisor node enabled
in dts. It is necessary action if we want to use driver for Xen
Initial domain shield, that means that the driver isn't tied only
to XenVM board configuration.

Add subdirectory of 'xen' when we have Xen Hypervisor node enabled
in dts. It is necessary action if we want to use arch arm64 core Xen
related code for Xen Initial domain shield, that means that the core
of Xen isn't tied only to XenVM SoC configuration.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
arch: arm64: mmu: WA: avoid using of set/way cache instructions for Xen

Unfortunately, set/way ops are not easily virtualized by Xen. S/W
emulation is disabled, because IP-MMU is active for Dom0. IP-MMU is
a IO-MMU made by Renesas, as any good IO-MMU, it shares page-tables
with CPU. Trying to emulate S/W with IP-MMU active will lead to
IO-MMU faults. So if we build Zephyr as a Xen Initial domain, it
won't work with cache management support enabled.

In this commit I use VA data invalidate instead of set/way --
invalidate all virtual address space, so it might increase system
initialization time.

Note: this patch is applied as a simple and fast workaround, we need
      to find more optimal solution.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
drivers: xen: gnttab: do Xen node mapping inside driver

Move memory mapping of Xen node to Grant Table driver system init
function. After moving mapping we don't need anymore records of
xen-xen node into 'mmu_regions' array, so they were deleted from
all SoCs: Rcar Gen3/Gen4 and XenVM.

We need at least 16M of virtual address space to map memory of Xen
node, so the virtual memory sized has been increased to 32 MB, it
should be enough for basic use-cases and mapping of 16M mem region
of Xen node.

Unfortunately, after moving we also need to increase number of XLAT
tables. The previous code was more efficient if we talking about
usage of XLAT tables, because it mapped grant tables using a higher-
order table that allows mapping blocks of 2MB. And after the changes
is maps every 4KB page, so we need more XLAT tables.

Note: Xen Grant Table driver doesn't use Zephyr Device Model.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop commented:
fix all Compliance Checks except new typedefs and license MIT

xakep-amatop commented on drivers/xen/dom0/CMakeLists.txt:
```diff
@@ -0,0 +1,4 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (c) 2021 EPAM Systems
```
change year here and in other files

xakep-amatop commented on boards/arm64/xenvm/board.cmake:
```diff
@@ -0,0 +1,14 @@
+set(SUPPORTED_EMU_PLATFORMS qemu)
```
looks like these changes should be dropped too

xakep-amatop commented on drivers/xen/memory.c:
```diff
@@ -0,0 +1,67 @@
+#include <zephyr/arch/arm64/hypercall.h>
```
add preamble

xakep-amatop commented on include/zephyr/xen/dom0/zimage.h:
```diff
@@ -0,0 +1,30 @@
+/*
```
(note) drop this file here and move it to zephyr zenlib 
@firscity @oleksiimoisieiev FYI

xakep-amatop commented on include/zephyr/xen/generic.h:
```diff
@@ -32,4 +32,10 @@
 				__ATOMIC_SEQ_CST) ? new : old; \
 })
 
+#define NR_MAGIC_PAGES 4
```
drop this commit and move changes to zephyr xen-lib
@firscity @oleksiimoisieiev

xakep-amatop commented on include/zephyr/xen/public/io/xenbus.h:
```diff
@@ -0,0 +1,73 @@
+/* SPDX-License-Identifier: MIT */
```
drop this commit here and move to xenstore srv app
@firscity @oleksiimoisieiev

xakep-amatop commented on include/zephyr/xen/dom0/uimage.h:
```diff
@@ -0,0 +1,31 @@
+/*
```
drop the file from zephyr and move this file to xen-lib 
@firscity @oleksiimoisieiev

xakep-amatop commented on dts/bindings/xen/xen,xen-4.15.yaml:
```diff
@@ -1,11 +1,5 @@
-description: Xen Platform Control Registers
+description: Xen Platform Control Registers v4.15
```
change version in all places to 4.17

xakep-amatop commented on include/zephyr/xen/public/arch-arm.h:
```diff
@@ -205,15 +205,138 @@ typedef uint64_t xen_pfn_t;
  */
 #define XEN_LEGACY_MAX_VCPUS 1
 
+typedef uint64_t xen_ulong_t;
```
check if these changes from xen 4.17 or not

xakep-amatop commented on arch/arm64/core/mmu.c:
```diff
@@ -820,7 +820,12 @@ static void enable_mmu_el1(struct arm_mmu_ptables *ptables, unsigned int flags)
 
 	/* Invalidate all data caches before enable them */
+#ifdef CONFIG_DT_HAS_XEN_XEN_ENABLED
+	/* todo: find more optimal solution */
```
can we merge it into mainline?..

firscity commented on include/zephyr/xen/public/domctl.h:
```diff
@@ -0,0 +1,387 @@
+/******************************************************************************
```
Add license identifier and fix code style in this file

firscity commented on include/zephyr/xen/public/domctl.h:
```diff
@@ -0,0 +1,387 @@
+#include "grant_table.h"
+#include "memory.h"
+
+#define XEN_DOMCTL_INTERFACE_VERSION 0x00000014
```
0x15 should be in initial commit, need to squash with some upper commit where it was made.

firscity commented on include/zephyr/xen/public/io/libxenvchan.h:
```diff
@@ -0,0 +1,88 @@
+/* SPDX-License-Identifier: MIT */
```
Move it to zephyr-xenlib

firscity commented on drivers/xen/CMakeLists.txt:
```diff
@@ -4,3 +4,4 @@
 zephyr_sources(hvm.c)
 zephyr_sources(events.c)
 zephyr_sources_ifdef(CONFIG_XEN_GRANT_TABLE gnttab.c)
+zephyr_sources(memory.c)
```
This change should be in separate commit

firscity commented on drivers/xen/memory.c:
```diff
@@ -0,0 +1,72 @@
+/*
```
These changes should be in separate commit

firscity commented on drivers/xen/memory.c:
```diff
@@ -0,0 +1,72 @@
+{
+	struct xen_add_to_physmap xatp;
+
+	memset(&xatp, 0, sizeof(xatp));
```
`xatp = {0}`

firscity commented on drivers/xen/memory.c:
```diff
@@ -0,0 +1,72 @@
+{
+	struct xen_add_to_physmap_batch xatpb;
+
+	memset(&xatpb, 0, sizeof(xatpb));
```
`xatpb = {0}`

firscity commented on drivers/xen/memory.c:
```diff
@@ -0,0 +1,72 @@
+{
+	struct xen_memory_reservation reservation;
+
+	memset(&reservation, 0, sizeof(reservation));
```
`reservation = {0}`

firscity commented on include/zephyr/xen/memory.h:
```diff
@@ -0,0 +1,21 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* TODO: fix headers including and add docs for functions */
```
Fix this todo before upstream

firscity commented on drivers/xen/dom0/domctl.c:
```diff
@@ -0,0 +1,322 @@
+{
+	xen_domctl_t domctl;
+
+	memset(&domctl, 0, sizeof(domctl));
```
`domctl = {0}` everywhere instead of memset

firscity commented on drivers/xen/dom0/domctl.c:
```diff
@@ -0,0 +1,322 @@
+		bind->u.spi.spi = spi;
+		break;
+	default:
+		/* TODO: check what to do with others */
```
Check this todo before upstream

firscity commented on include/zephyr/xen/public/domctl.h:
```diff
@@ -37,7 +37,7 @@
 #include "memory.h"
 
-#define XEN_DOMCTL_INTERFACE_VERSION 0x00000014
+#define XEN_DOMCTL_INTERFACE_VERSION 0x00000015
```
move to initial commit

firscity commented on include/zephyr/xen/public/domctl.h:
```diff
@@ -69,9 +69,11 @@ struct xen_domctl_createdomain {
 #define _XEN_DOMCTL_CDF_nested_virt   6
 #define XEN_DOMCTL_CDF_nested_virt    (1U << _XEN_DOMCTL_CDF_nested_virt)
+/* Should we expose the vPMU to the guest? */
+#define XEN_DOMCTL_CDF_vpmu           (1U << 7)
```
move all changes to public header to initial commit

firscity commented on dts/bindings/xen/xen,xen-4.15.yaml:
```diff
@@ -1,11 +1,5 @@
-description: Xen Platform Control Registers
+description: Xen Platform Control Registers v4.15
 
 compatible: "xen,xen-4.15"
```
Probably we can drop exact version at all, just use "xen,xen"

xakep-amatop replied:
yes we can

firscity commented on boards/arm64/xenvm/xenvm_gicv3_defconfig:
```diff
@@ -7,7 +7,7 @@ CONFIG_XENVM_USE_GIC_V3=y
 CONFIG_SERIAL=y
 
-CONFIG_MAX_XLAT_TABLES=10
+CONFIG_MAX_XLAT_TABLES=24
```
Same should be done for `xenvm_defconfig`
FYI: @xakep-amatop

firscity committed:
xen: implement Xen domain control interface

Add Xen domctl API implementation for Zephyr as control domain.

Previously Zephyr OS was used as unprivileged Xen domain (Domain-U),
but it also can be used as lightweight Xen control domain (Domain-0).
To implement such fuctionality additional Xen interfaces are needed.
One of them is Xen domain controls (domctls) - it allows to create,
configure and manage Xen domains.

Also, used it as a possibility to update files copyright and licenses
identifiers in touched files.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

firscity committed:
xen: add helper functions for Xen domain memory management

Add wrapper functions for Xen memory managment hypercall. They can be
used for unprivilaged Zephyr guest initialization and for domain
management, when Zephyr is used as Xen Domain 0.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop commented on drivers/xen/dom0/domctl.c:
```diff
@@ -0,0 +1,302 @@
+
+int xen_domctl_scheduler_op(int domid, struct xen_domctl_scheduler_op *sched_op)
+{
+	xen_domctl_t domctl = {0};
```
(note) you can use the next initializer here and below:
```
xen_domctl_t domctl = {
	.cmd = XEN_DOMCTL_scheduler_op,
	.domain = domid,
	.u.scheduler_op = *sched_op,
};
```
and rest of fields automatically will be initialized

firscity replied:
@xakep-amatop thanks for suggestion, done.

xakep-amatop commented:
note: not needed anymore --- clear changes you can find here: https://github.com/xen-troops/zephyr/pull/37
